### ISSUE TITLE
Problems when using Octant with multiple kubeconfig

### ISSUE THREAD

**Issue by jungopro** [vmware-tanzu/octant#518]
Hello

This is similar to #147 but I don't think this is resolved
I have multiple kubeconfig files that I'm using through the **KUBECONFIG** env var

**What I expected?**

- By default, running octant with no params will auto select the active context from the merged KUBECONFIG env var
- Using octant with --kubeconfig param will work regardless of the KUBECONFIG current values

**What actually happened?**

The behaviour isn't consistent and I couldn't pin-point the exact time when this happens. See my try-and-error below (i renamed the yamls for readability)

```console
# setting KUBECONFIG to a single file, cluster-a
─ KUBECONFIG=/Users/omerbarel/.kube/cluster-a.yaml

# running octant works as expected
─ octant
2020-01-06T10:33:59.303+0200	INFO	module/manager.go:79	registering action	{"component": "module-manager", "actionPath": "overview/serviceEditor", "module-name": "overview"}
2020-01-06T10:33:59.309+0200	INFO	module/manager.go:79	registering action	{"component": "module-manager", "actionPath": "deployment/configuration", "module-name": "overview"}
2020-01-06T10:33:59.309+0200	INFO	module/manager.go:79	registering action	{"component": "module-manager", "actionPath": "overview/containerEditor", "module-name": "overview"}
2020-01-06T10:33:59.309+0200	INFO	module/manager.go:79	registering action	{"component": "module-manager", "actionPath": "octant/deleteObject", "module-name": "configuration"}
2020-01-06T10:33:59.312+0200	INFO	dash/dash.go:370	Using embedded Octant frontend
2020-01-06T10:33:59.324+0200	INFO	dash/dash.go:349	Dashboard is available at http://127.0.0.1:7777

^C2020-01-06T10:34:25.045+0200	INFO	cluster/cluster.go:137	removing cluster client temporary directory	{"component": "cluster client"}
2020-01-06T10:34:25.046+0200	INFO	plugin/manager.go:317	shutting down plugin watcher

# running octant with the same file cluster-a via the --kubeconfig param also works:

─ octant --kubeconfig ~/.kube/cluster-a.yaml
2020-01-06T10:34:43.700+0200	INFO	module/manager.go:79	registering action	{"component": "module-manager", "actionPath": "deployment/configuration", "module-name": "overview"}
2020-01-06T10:34:43.701+0200	INFO	module/manager.go:79	registering action	{"component": "module-manager", "actionPath": "overview/containerEditor", "module-name": "overview"}
2020-01-06T10:34:43.701+0200	INFO	module/manager.go:79	registering action	{"component": "module-manager", "actionPath": "overview/serviceEditor", "module-name": "overview"}
2020-01-06T10:34:43.701+0200	INFO	module/manager.go:79	registering action	{"component": "module-manager", "actionPath": "octant/deleteObject", "module-name": "configuration"}
2020-01-06T10:34:43.701+0200	INFO	dash/dash.go:370	Using embedded Octant frontend
2020-01-06T10:34:43.714+0200	INFO	dash/dash.go:349	Dashboard is available at http://127.0.0.1:7777

^C2020-01-06T10:34:55.478+0200	INFO	cluster/cluster.go:137	removing cluster client temporary directory	{"component": "cluster client"}

# setting KUBECONFIG to a single file, cluster-b

─ KUBECONFIG=/Users/omerbarel/.kube/cluster-b.yaml

# running octant works as expected

─ octant
2020-01-06T10:35:17.162+0200	INFO	module/manager.go:79	registering action	{"component": "module-manager", "actionPath": "overview/serviceEditor", "module-name": "overview"}
2020-01-06T10:35:17.162+0200	INFO	module/manager.go:79	registering action	{"component": "module-manager", "actionPath": "deployment/configuration", "module-name": "overview"}
2020-01-06T10:35:17.162+0200	INFO	module/manager.go:79	registering action	{"component": "module-manager", "actionPath": "overview/containerEditor", "module-name": "overview"}
2020-01-06T10:35:17.162+0200	INFO	module/manager.go:79	registering action	{"component": "module-manager", "actionPath": "octant/deleteObject", "module-name": "configuration"}
2020-01-06T10:35:17.165+0200	INFO	dash/dash.go:370	Using embedded Octant frontend
2020-01-06T10:35:17.173+0200	INFO	dash/dash.go:349	Dashboard is available at http://127.0.0.1:7777

^C2020-01-06T10:35:31.973+0200	INFO	plugin/manager.go:317	shutting down plugin watcher
2020-01-06T10:35:31.973+0200	INFO	cluster/cluster.go:137	removing cluster client temporary directory	{"component": "cluster client"}

# running octant with the same file cluster-b via the --kubeconfig param also works:

─ octant --kubeconfig ~/.kube/cluster-b.yaml
2020-01-06T10:35:52.655+0200	INFO	module/manager.go:79	registering action	{"component": "module-manager", "actionPath": "overview/serviceEditor", "module-name": "overview"}
2020-01-06T10:35:52.656+0200	INFO	module/manager.go:79	registering action	{"component": "module-manager", "actionPath": "deployment/configuration", "module-name": "overview"}
2020-01-06T10:35:52.656+0200	INFO	module/manager.go:79	registering action	{"component": "module-manager", "actionPath": "overview/containerEditor", "module-name": "overview"}
2020-01-06T10:35:52.656+0200	INFO	module/manager.go:79	registering action	{"component": "module-manager", "actionPath": "octant/deleteObject", "module-name": "configuration"}
2020-01-06T10:35:52.656+0200	INFO	dash/dash.go:370	Using embedded Octant frontend
2020-01-06T10:35:52.665+0200	INFO	dash/dash.go:349	Dashboard is available at http://127.0.0.1:7777

^C2020-01-06T10:36:04.753+0200	INFO	cluster/cluster.go:137	removing cluster client temporary directory	{"component": "cluster client"}

# KUBECONFIG is still set to cluster-b
# passing cluster-a as parameter value, also works:

─ octant --kubeconfig ~/.kube/cluster-a.yaml
2020-01-06T10:36:18.340+0200	INFO	module/manager.go:79	registering action	{"component": "module-manager", "actionPath": "deployment/configuration", "module-name": "overview"}
2020-01-06T10:36:18.341+0200	INFO	module/manager.go:79	registering action	{"component": "module-manager", "actionPath": "overview/containerEditor", "module-name": "overview"}
2020-01-06T10:36:18.341+0200	INFO	module/manager.go:79	registering action	{"component": "module-manager", "actionPath": "overview/serviceEditor", "module-name": "overview"}
2020-01-06T10:36:18.341+0200	INFO	module/manager.go:79	registering action	{"component": "module-manager", "actionPath": "octant/deleteObject", "module-name": "configuration"}
2020-01-06T10:36:18.342+0200	INFO	dash/dash.go:370	Using embedded Octant frontend
2020-01-06T10:36:18.349+0200	INFO	dash/dash.go:349	Dashboard is available at http://127.0.0.1:7777

^C2020-01-06T10:36:26.929+0200	INFO	cluster/cluster.go:137	removing cluster client temporary directory	{"component": "cluster client"}
2020-01-06T10:36:26.930+0200	INFO	plugin/manager.go:317	shutting down plugin watcher

# creating a merged KUBECONFIG env:

KUBECONFIG=/Users/omerbarel/.kube/cluster-b.yaml:/Users/omerbarel/.kube/cluster-a.yaml

# running octant still works, cluster b was selected automatically:

─ kubectx
cluster-a
*cluster-b*

─ octant
2020-01-06T10:37:01.740+0200	INFO	module/manager.go:79	registering action	{"component": "module-manager", "actionPath": "deployment/configuration", "module-name": "overview"}
2020-01-06T10:37:01.741+0200	INFO	module/manager.go:79	registering action	{"component": "module-manager", "actionPath": "overview/containerEditor", "module-name": "overview"}
2020-01-06T10:37:01.741+0200	INFO	module/manager.go:79	registering action	{"component": "module-manager", "actionPath": "overview/serviceEditor", "module-name": "overview"}
2020-01-06T10:37:01.741+0200	INFO	module/manager.go:79	registering action	{"component": "module-manager", "actionPath": "octant/deleteObject", "module-name": "configuration"}
2020-01-06T10:37:01.743+0200	INFO	dash/dash.go:370	Using embedded Octant frontend
2020-01-06T10:37:01.752+0200	INFO	dash/dash.go:349	Dashboard is available at http://127.0.0.1:7777

^C2020-01-06T10:37:07.408+0200	INFO	cluster/cluster.go:137	removing cluster client temporary directory	{"component": "cluster client"}
2020-01-06T10:37:07.408+0200	INFO	plugin/manager.go:317	shutting down plugin watcher

# switching context

─ kubectx cluster-a
Switched to context "cluster-a.

# running octant fails

─ octant
2020-01-06T10:37:15.157+0200	ERROR	commands/dash.go:96	dashboard failed	{"err": "failed to init cluster client: invalid configuration: [context was not found for specified context: cluster-a, cluster has no server defined]"}
github.com/vmware-tanzu/octant/internal/commands.newOctantCmd.func1.2
	/private/tmp/octant-20191108-95744-1plrk02/src/github.com/vmware-tanzu/octant/internal/commands/dash.go:96
```

**Comment by jungopro** [comment]
**Recommendation**: maybe allow running octant with **--noconfig** and simply allow adding the kubeconfig via copy-paste in the GUI?
This is a bit similar to the way **Kontena Lens** operates

**Comment by wwitzel3** [comment]
Thank you for the great reproduction steps. I will try to look at this. We do have an issue for allowing Octant to start without a kubeconfig #362 I think extending that to include the scenario where Octant fails to start because for whatever reason it was unable to load the active context would be a better experience as long as the error message was presented in the UI.

**Comment by mattouille** [comment]
Here's my use case/workflow. The disclaimer being that I don't know if there's SIG's or WG's trying to solve this kind of issue too (or standardize how this is solved?)

**Storing KubeConfig**
I have to access multiple clusters (hundreds) and thus I now have a directory called `$HOME/.kube/configs` (not to be confused with `$HOME/.kube/config`). In this directory is all the kube configs for various clusters.

**Referencing KubeConfig**
In `$HOME/.zshenv` I have some code that iterates that directory and does a little `export KUBECONFIG=$KUBECONFIG:<new_path>`. This allows me to use kubectx/kubens to switch between kubeconfigs quickly.

**Comment by totallyGreg** [comment]
Octant seems to have trouble if a colon ':' is present in my KUBECONFIG.  And presents the "upload config file" window on launch.  
I also set my KUBECONFIG like a PATH with my `~/.kube/config` first and then a set of directories.  
When no directories are present my KUBECONFIG was listed as `/Users/username/.kube/confg:`
If I remove the colon at the end, Octant launches fine without asking for a kubeconfig file.

**Comment by jldec** [comment]
I'm on a Mac with 2 kubeconfigs, the current-context is in the 2nd file in $KUBECONFIG.
Octant crashes a second after showing the visual prompt for kubeconfig in the browser.

```
$ octant version
Version:  0.14.1
Git commit:  51ce7f6a7e87c9cfcabc9bf402dff7eda5d6986b
Built:  2020-07-20T23:52:37Z

$ echo $KUBECONFIG
/Users/jldec/.kube/config:/Users/jldec/tss/.kubeconfig

$ kubectl config get-contexts
CURRENT   NAME                 CLUSTER              AUTHINFO                  NAMESPACE
*         decc                 decc-scdc1-staging   scdc1-staging-tss-admin   scdc1-staging-tss
          docker-desktop       docker-desktop       docker-desktop            
          docker-for-desktop   docker-desktop       docker-desktop            

$ octant
2020-07-24T10:01:16.801+0100	INFO	dash/dash.go:105	no valid kube config found, initializing loading API
2020-07-24T10:01:16.801+0100	INFO	dash/dash.go:140	waiting for kube config ...
2020-07-24T10:01:16.805+0100	INFO	dash/dash.go:497	Dashboard is available at http://127.0.0.1:7777

2020-07-24T10:01:18.614+0100	ERROR	dash/dash.go:557	cannot create api: failed to init cluster client, does your kube config have a current-context set?: invalid configuration: [context was not found for specified context: decc, cluster has no server defined]
github.com/vmware-tanzu/octant/pkg/dash.(*Runner).startAPIService
	/__w/octant/octant/pkg/dash/dash.go:557
github.com/vmware-tanzu/octant/pkg/dash.(*Runner).Start.func2
	/__w/octant/octant/pkg/dash/dash.go:146
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x40 pc=0x23750c6]

goroutine 39 [running]:
github.com/vmware-tanzu/octant/internal/api.(*API).Handler(0x0, 0x49dbe80, 0xc00019e8d0, 0xc0, 0x278fe60, 0x27b7f01, 0xc0001d9200)
	/__w/octant/octant/internal/api/api.go:142 +0x26
github.com/vmware-tanzu/octant/pkg/octant.(*HandlerFactory).Handler(0xc00154b410, 0x49dbe80, 0xc00019e8d0, 0x4789d68, 0xc00154f620, 0x1, 0xc00005800b)
	/__w/octant/octant/pkg/octant/handler.go:58 +0x98
github.com/vmware-tanzu/octant/pkg/dash.(*Runner).startAPIService(0xc00019ebd0, 0x49dbe80, 0xc00019e8d0, 0x49efc20, 0xc000784308, 0x0, 0xc00005800b, 0x1d, 0x0, 0x0, ...)
	/__w/octant/octant/pkg/dash/dash.go:568 +0x1cb
github.com/vmware-tanzu/octant/pkg/dash.(*Runner).Start.func2(0xc00019ebd0, 0x49efc20, 0xc000784308, 0xc000158000, 0xc000150000, 0x49dbe80, 0xc00019e8d0)
	/__w/octant/octant/pkg/dash/dash.go:146 +0x1c2
created by github.com/vmware-tanzu/octant/pkg/dash.(*Runner).Start
	/__w/octant/octant/pkg/dash/dash.go:137 +0x379
```
